cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -292,7 +292,11 @@ withContextAndSelectors
   -> IO b
 withContextAndSelectors noTargets kind flags@NixStyleFlags{..} targetStrings globalFlags cmd act =
   withTemporaryTempDirectory $ \mkTmpDir -> do
-    (tc, ctx) <- withProjectOrGlobalConfig verbosity ignoreProject globalConfigFlag withProject (withoutProject mkTmpDir)
+    (tc, ctx) <-
+      withProjectOrGlobalConfig
+        ignoreProject
+        withProject
+        (withGlobalConfig verbosity globalConfigFlag $ withoutProject mkTmpDir)
```

Comment:
Tbh I dislike that `withProjectOrGlobalConfig` will not apply any "global config" if you do not remember to do `withGlobalConfig` at every call site (and there are still a few call sites where we add `withGlobalConfig`). Although the function now is much better...

Reply:
I hear you. I think that function is terrible, all it does is unconditionally run `with` and then run `without` if `with` throws an exception.
This is actually hilarious: because `withoutProject` starts with:
```
withoutProject _ | null targetStrings = withProject
```
Running `cabal install` in a non project setting will: invoke withProject, catch the exception, invoke withoutProject, which calls withProject again, raising a second exception.

Can't we just check for `cabal.project` or `*.cabal`? :shrug: another time ...

With respect to the global config; happy to expore other ways but so far I made them equal (in rebuildProjectConfig there is a fun dance with cliConfig).

Reply:
> Running cabal install in a non project setting will: invoke withProject, catch the exception, invoke withoutProject, which calls withProject again, raising a second exception.

Sigh...